### ISSUE TITLE
OpenCV as alternative library for map.image.affine_transform

### ISSUE THREAD

**Issue by cbard** [sunpy/sunpy#4478]
### Description

As part of the [Helio Hackweek](https://github.com/heliohackweek/AIA_acceleration) a few weeks ago, Raphael Attie (@WaaallEEE) and I experimented with using the OpenCV python library, specifically `cv2.warpAffine`, as an alternative to `scipy.ndimage.transform` and `skimage.warp`. I have now figured out the nuances regarding different coordinate systems and order of operations, and have been able to essentially reproduce the current behavior of `map.rotate` and `image.affine_transform` using `cv2.warpAffine` as the underlying image manipulator.

I've added my [changes to a local fork](https://github.com/cbard/sunpy/blob/opencv2_warp/sunpy/image/transform.py#L123) and run [some benchmarking tests](https://github.com/cbard/sunpy/blob/opencv2_warp/open_cv_tests/CV_rotate_tests2.ipynb). Basically, `openCV.warpAffine` is 3x faster than `skimage.warp` on my laptop.

I have not created a pull request, since discussions with @wtbarnes and @Cadair suggest that it may be better to add optional argument(s) in `image.affine_transform` and `map.rotate` that allow for the passing in of custom `affine_transform` functions and their related `**kwargs`. My current fork has the openCV implementation hardcoded, so I wanted to get some information on the best way to proceed.

A difficulty with passing through a general function (+arguments) is that the required arguments for the underlying image manip routines are calculated within `image.affine_transform` or `map.rotate`; i.e. they are hidden from the general user. Even passing in an angle or rotation matrix to `map.rotate` may require some additional processing to calculate a rotation matrix and put it in the right format for the underlying routine. It seems weird to me to have the user do the required calculations and then pass the resulting arguments through `rotate()`. That kinda defeats the purpose of having `rotate()` in the first place.

**Comment by dstansby** [comment]
Personally I think it would be reasonable to add a `method=` argument to `rotate()`, that would replace the `use_scipy` argument, and would then allow values from `'scipy', 'skimage', 'opencv'`. Then code can be added to sunpy to deal with the opencv rotation, and users only need to have opencv installed if they want to be able to use the 'opencv' option.

**Comment by cbard** [comment]
Yes, I think that's a better suggestion. I believe it's easy enough to change the method handling in `affine_transform()` without touching `rotate()` (except to add the `method=` argument).

**Comment by wtbarnes** [comment]
I think the larger issue here though is whether we want to have a few hard-coded options (as we do now) or if we want to allow for arbitrary functions to be passed in to handle the affine transform. I think ideally we would have the latter and then maybe `sunpy` would include options for using scipy or scikit-image as we already do, with these specified as a string. So maybe this would look something like,

```python
def affine_transform(..., method='scipy', ...):
    if isinstance(method, str):
        # Check whether this is scipy or scikit-image
        # Set method to the appropriate method for whichever method is chosen
    if not isinstance(method, types.FunctionType):
        # At this point, method has to be a function
        # Otherwise, throw an error
    # Continue with the logic of the affine transform, using method to do the transformation    
```

I'm just weary of maintaining two many different options ourselves as well as further complicating the logic in `affine_transform`. If we provide this type of entrypoint, it is then easier for downstream packages (e.g. `aiapy`) to then shove in whatever type of transform implementation they like (e.g. OpenCV) while still leveraging the `Map.rotate` API. Also, at some point, there's bound to be some other, faster, fancier affine transform method that someone will want to try and this helps future proof us a bit.

**Comment by cbard** [comment]
> I think the larger issue here though is whether we want to have a few hard-coded options (as we do now) or if we want to allow for arbitrary functions to be passed in to handle the affine transform.

Such decisions are above my paygrade, but I have found it easy to leverage the pre- and post-processing in `rotate()`  and `image.affine_transform()` if we define the library-specific routines relative to some base function (`scipy.ndimage.affine_transform`) like so:
```python
def _scipy_affine_transform(image,rmatrix,order,scale,missing,shift):
    # do nothing (except maybe set NaNs to zero)
    return scipy.ndimage.affine_transform(arguments)

def _opencv_affine_transform(image,rmatrix,order,scale,missing,shift):
    #do preprocessing of input arguments for openCV 
    # (e.g. calculate rotation matrix, convert image to float)
    return cv2.warpAffine(arguments)

def _skimage_affine_transform(image,rmatrix,order,scale,missing,shift):
    # do preprocessing of input arguments for skimage
    # (e.g. calculate rotation matrix, convert image to float, scale/rescale image?)
    return skimage.transform.warp(arguments)
```
and the main `image.affine_transform()` would be something like
```python
def affine_transform(args, method='scipy'):
    # preprocessing of `shift` same as current
    if method == 'skimage':
       return _skimage_affine_transform(image,rmatrix,order,scale,missing,shift)
    elif method == 'opencv':
       return _opencv_affine_transform(image,rmatrix,order,scale,missing,shift)
    else:
       return _scipy_affine_transform(image,rmatrix,order,scale,missing,shift)
```
This would make it easy to manage the logic of switching between different hardcoded library implementations (and add new ones in the future). This would also make it easy to adapt for drop-in functions (with switching logic as @wtbarnes has provided above), provided that the drop-in functions use the required arguments (and no other): `image,rmatrix,order,scale,missing,shift` .

**Comment by dstansby** [comment]
We had a chat about this on our weekly dev call, and and decided the best way forward with this is to accept an arbitrary function as the `method` argument. I think the next step here (open to being worked on to whoever wants to) is to define what the function signature we accept is going to be.

**Comment by cbard** [comment]
Ok, sounds good. I'm 80% of the way to having drop-in arbitrary functions in `rotate()`, so I'll make a pull request in the next few days. 

It looks like the drop-in function signature is probably going to be the same as the current implementation of `affine_transform`, except without the `method=` argument. I think openCV is different enough from skimage and scipy.ndimage that the rotation, scaling, and translation will have to be calculated within the drop-in function.

**Comment by nabobalis** [comment]
Updated version of this was merged in https://github.com/sunpy/sunpy/pull/6089 This will be available in sunpy 4.0
